pdpctrl-client/host.go

```diff
@@ -11,57 +13,88 @@ import (
 	pb "github.com/infobloxopen/themis/pdp-control"
 )
 
+type Logger interface {
+	Infof(format string, a ...interface{})
+	Errorf(format string, a ...interface{})
+}
+
 type Host struct {
 	address string
 	chunk   int
 
 	conn   *grpc.ClientConn
-	client *pb.PDPControlClient
+	client pb.PDPControlClient
 
 	ready  bool
 	policy int32
+
+	log Logger
+}
+
+type VersionError struct {
+	version string
+}
+
+func (e *VersionError) Error() string {
+	return e.version
 }
 
-func NewHost(addr string, chunk int) *Host {
-	return &Host{addr, chunk, nil, nil, false, -1}
+func NewHost(addr string, chunk int, log Logger) *Host {
+	return &Host{
+		address: addr,
+		chunk:   chunk,
+		policy:  -1,
+		log:     log,
+	}
 }
 
-func (h *Host) connect(timeout time.Duration, log Logger) error {
-	log.Infof("Connecting to %s...", h.address)
+func (h *Host) Connect(timeout time.Duration) error {
+	h.log.Infof("Connecting to %s...", h.address)
 	conn, err := grpc.Dial(h.address, grpc.WithInsecure(), grpc.WithBlock(), grpc.WithTimeout(timeout))
 	if err != nil {
-		log.Errorf("Can't connect to %s: %v. Skipping...", h.address, err)
+		h.log.Errorf("Can't connect to %s: %v. Skipping...", h.address, err)
 		return err
 	}
 
 	h.conn = conn
+	h.client = pb.NewPDPControlClient(h.conn)
 
-	client := pb.NewPDPControlClient(h.conn)
-	h.client = &client
 	return nil
 }
 
-func (h *Host) close() {
+func (h *Host) Close() {
 	if h.conn != nil {
 		h.conn.Close()
 	}
 
 	h.client = nil
 }
 
-func (h *Host) uploadIncludes(includes map[string][]byte, log Logger) []int32 {
+func (h *Host) Process(patch bool, version string, includes map[string][]byte, policy []byte) error {
```

Comment:
>The main part of cluster abstraction was to separate upload and apply into two phases on whole set of servers

Why do we need this property? If some uploads will fail we end up with partially updated PDPs anyway. From the other hand independent PDP management should simplify our logic